[ty] route constructor calls through bindings

Hugo-Polloli opened the pull request:
## Summary

fixes https://github.com/astral-sh/ty/issues/2124

* route constructor calls through the existing `Type::bindings` machinery via `constructor_bindings()`, ensuring constructor return types are instance-typed (not `__init__`'s `None`)
* delete the legacy constructor call path
* constructor overload resolution is deterministic (first matching overload), and implicit `__new__`/`__init__` lints remain method-specific and fire once per call site (including the missing-`__init__` fallback)

## Test Plan

* mdtest (constructor): added coverage for `@staticmethod __new__`, conflicting `__new__`/`__init__` parameter types, `__new__`-only fallback (object `__init__` accepts args), and generic constructor inference.
* mdtest (decorators): `functools.cached_property` specifically added to address https://github.com/astral-sh/ty/issues/1446
* mdtest (classes): fixed the TODOs for generic class params `C[Unknown]` are now correctly revealed as `C[int]` (and constructor specialization uses first matching overload)
* mdtest (functions): improved constructor path now correctly surfaces both the ill-typed generic call and the resulting invalid assignment

carljm commented:
Wow, awesome, thank you for working on this!

It looks like the diagnostics that this PR removes on lines 81 and 82 of the conformance suite, in `constructors_call_type.py`, are supposed to be there, so that looks like a regression in this PR (and maybe some coverage we are missing in mdtests.)

The diagnostic added on line 29 of `generics_scoping.py` is correct, looks like a fix from this PR!

Hugo-Polloli commented on crates/ty_python_semantic/resources/mdtest/generics/pep695/functions.md:
```diff
@@ -941,7 +941,9 @@ TMsg = TypeVar("TMsg", bound=Msg)
 class Builder(Generic[TMsg]):
     def build(self) -> Stream[TMsg]:
         stream: Stream[TMsg] = Stream()
-        # TODO: no error
```
the previous TODO mentioned no error but to my knowledge the comment I added is correct and the two errors surfaced should indeed appear, so I'd like your input on which it should be, thanks !

Hugo-Polloli commented:
> Wow, awesome, thank you for working on this!
> 
> It looks like the diagnostics that this PR removes on lines 81 and 82 of the conformance suite, in `constructors_call_type.py`, are supposed to be there, so that looks like a regression in this PR (and maybe some coverage we are missing in mdtests.)
> 
> The diagnostic added on line 29 of `generics_scoping.py` is correct, looks like a fix from this PR!

Indeed looks like a regression ! That was because I missed updating the TypeVar matching arms, so thanks for catching that, I fixed it to validate params against the bound class and added an mdtest (straight copied from typing conformance) to prevent future regression on that side

CI is currently red for reasons that look unrelated to this change: mypy_primer is failing while cloning pypa/twine (git clone … exit 128), and codspeed is panicking. I’ll rerun CI and dig in after Christmas if issues persist :)

sinon commented:
> CI is currently red for reasons that look unrelated to this change: mypy_primer is failing while cloning pypa/twine (git clone … exit 128), and codspeed is panicking. I’ll rerun CI and dig in after Christmas if issues persist :)

The CodSpeed panic is related to your change but not necessarily a blocker to the change. There is an assertion in each benchmarked library that the number of diagnostics are >0 but less than a per-library upperbound but your change means it is now above this bound. If the new diagnostics are not false-positives then you just increase the number in the relevant file (plus some headroom). The relevant setting is here: https://github.com/astral-sh/ruff/blob/5ea30c4c53894dc35fa5af2ca84757ec24f3f7d0/crates/ruff_benchmark/benches/ty_walltime.rs#L197

Hugo-Polloli commented:
> > CI is currently red for reasons that look unrelated to this change: mypy_primer is failing while cloning pypa/twine (git clone … exit 128), and codspeed is panicking. I’ll rerun CI and dig in after Christmas if issues persist :)
> 
> The CodSpeed panic is related to your change but not necessarily a blocker to the change. There is an assertion in each benchmarked library that the number of diagnostics are >0 but less than a per-library upperbound but your change means it is now above this bound. If the new diagnostics are not false-positives then you just increase the number in the relevant file (plus some headroom). The relevant setting is here:
> 
> https://github.com/astral-sh/ruff/blob/5ea30c4c53894dc35fa5af2ca84757ec24f3f7d0/crates/ruff_benchmark/benches/ty_walltime.rs#L197

Thanks for the explanation ! I found the way to run it locally, I'll check the diffs and whether they're regressions/bugs when I get back :)

Hugo-Polloli commented:
So I checked the new sympy diagnostics, they come from strict unions of constructors and `type[Self]`. I added corresponding mdtests and TODOs, since some of these can currently be considered false positives, but properly fixing them would require additional modelling (runtime-config constructor aliases, subclass-widened constructors etc...)
I bumped the max diagnostics for sympy

Hugo-Polloli commented:
The mypy_primer “duplicates” at `lib/spack/spack/cmd/create.py:1084:29` are from a call on a `Union[type[...]]` (24 variants). `report_diagnostics` currently emits one diagnostic per union element, and with `--output-format concise` the union subdiagnostics aren’t shown so they appear as identical repeated errors.

I found an existing TODO on `report_diagnostics` about aggregating union/overload failures into subdiagnostics, but implementing proper aggregation seems non-trivial and would probably be scope-creep for this PR.

I see two possibilities:
1. we accept the verbose/duplicated-looking output for union calls, follow up later with union-aware diagnostic aggregation (per the TODO).

2. some kind of minimal mitigation in `Bindings::report_diagnostics`, emit only one diagnostic for unions (first failing binding?) and add a subdiagnostic like “omitted N additional union variants”, quicker to do than proper aggregation

MichaReiser commented:
Thank you for working on this. Almost the entire team is out this week. It may take a few days before someone finds time to review your PR. Happy holidays.

Hugo-Polloli commented:
I see #22377 has been merged, I will work on starting the rebase work soon :)

carljm commented on crates/ty_python_semantic/resources/mdtest/call/constructor.md:
````diff
@@ -130,6 +149,30 @@ reveal_type(Foo(1))  # revealed: Foo
 ```
 
+## `__new__` defined as a staticmethod
+
````
```suggestion

`__new__` is implicitly a static method, but explicitly marking it as one is harmless:

```

carljm commented on crates/ty_python_semantic/resources/mdtest/type_of/generics.md:
```diff
@@ -141,6 +141,23 @@ class A:
+        # TODO: This is stricter than ideal for `type[Self]`; relaxing it safely would require
+        # modeling subclass-widened constructors and decorator-generated `__init__` (attrs, etc.).
+        # error: [invalid-argument-type] "Argument to bound method `__init__` is incorrect: Expected `int`, found `Literal["x"]`"
+        return self.__class__("x")
```
Add success case also, and verify the return type is correct?
```suggestion
    def clone(self: Self) -> Self:
        # error: [invalid-argument-type] "Argument to bound method `__init__` is incorrect: Expected `int`, found `Literal["x"]`"
        self.__class__("x")
        return self.__class__(1)
```

carljm commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -4031,11 +4031,11 @@ impl<'db> Type<'db> {
                 // their return type beyond what typeshed provides (though this support could
-                // likely be moved into the `try_call_constructor` path). Once we support
+                // likely be moved into the `constructor_bindings` path). Once we support
                 // overloads, re-evaluate the need for these arms.
```
We've supported overloads for a long time now :) It would be a great follow-up to this PR to see which of these arms can be removed now.

carljm commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -4361,20 +4361,22 @@ impl<'db> Type<'db> {
+                        ),
+                    )
+                    .into()
+                }),
```
Are there any `ClassLiteral` types for which `constructor_bindings` will return `None`? If so, why? If not, can this be just `.expect("all ClassLiteral types have a constructor signature")`?

It seems unclear after this PR why this fallback to the gradual signature even needs to exist.

carljm commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -4414,27 +4416,33 @@ impl<'db> Type<'db> {
+                // `__init__` signatures? and respect metaclass `__call__`).
+                SubclassOfInner::Class(class) => self
+                    .constructor_bindings(db)
+                    .unwrap_or_else(|| Type::from(class).bindings(db)),
```
Similar to above -- will this ever be `None`?

Hugo-Polloli replied:
Fixed that in the refactor: `constructor_bindings` now take `ClassType` and always returns `Bindings`

carljm commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -4414,27 +4416,33 @@ impl<'db> Type<'db> {
+                        ),
+                    )
+                    .into()
+                }),
```
Similar to above -- in what cases do we actually need this fallback anymore? Is it possible for a `SubclassOfInner::TypeVar` to have an inner type that isn't a class?

Ideally I feel like we could collapse the entire `GenericAlias` and `SubclassOf` branches into just one match that unconditionally defers to `constructor_bindings`. (Ideally the `ClassLiteral` branch too, but that requires getting rid of the special cases in that branch, too.)

carljm reviewed (changes requested):
This is awesome! Sorry you had to wait so long for review; we've been behind on reviews ever since the beta and holidays.

Left some inline comments, but I think the most important one is that I think we should adjust the signature of `constructor_bindings` and how it is called from `Type::bindings` to make it less error-prone: make it not a method on `Type` and not accept a `Type`, but rather accept just a `ClassLiteral`, and never return `None`.

Hugo-Polloli commented:
> This is awesome! Sorry you had to wait so long for review; we've been behind on reviews ever since the beta and holidays.
> 
> Left some inline comments, but I think the most important one is that I think we should adjust the signature of `constructor_bindings` and how it is called from `Type::bindings` to make it less error-prone: make it not a method on `Type` and not accept a `Type`, but rather accept just a `ClassLiteral`, and never return `None`.

Thanks a ton for the thorough review! No worries at all about the time to review!
I've addressed all your feedback throughout my last 3 commits,
[[ty] address review for mdtests and comments](https://github.com/astral-sh/ruff/pull/22124/changes/813bed8f855476afb0b0dbd28690f0427aad9bb3) for all the mdtests/comments suggestions
[[ty] create a known_class_literal_bindings helper](https://github.com/astral-sh/ruff/pull/22124/changes/271ff5822faabd4e189b7db0feebf4d62b4f7b95) is a transitory commit to move KnownClass constructor special‑cases into a helper so the later refactor stays reviewable and "diff‑light"
[[ty] refactor constructor_bindings to take ClassType](https://github.com/astral-sh/ruff/pull/22124/changes/e32c2119f5af395bbe12ebce03a3fe2a5cf0fa0a) is the main refactor: class‑like call handling is centralized, `constructor_bindings` now takes a `ClassType` and returns `Bindings`, the ClassLiteral/GenericAlias/SubclassOf paths are collapsed, and `type[T]` constraints keep their gradual behavior

I also preserved existing behavior around type[T] constraints and TypedDict/known‑class special cases, and added a couple clarifying comments where behavior could look surprising

Hugo-Polloli commented on crates/ty_python_semantic/src/types.rs:
```diff
@@ -4544,6 +4552,262 @@ impl<'db> Type<'db> {
+            Type::ClassLiteral(class) => (ClassType::NonGeneric(class), false),
+            Type::GenericAlias(alias) => (ClassType::Generic(alias), false),
+            Type::SubclassOf(subclass_of_type) => match subclass_of_type.subclass_of() {
+                SubclassOfInner::Dynamic(_) => return None,
```
Done, the dynamic `SubclassOf` path now directly produces a dynamic signature :)

carljm commented on crates/ty_python_semantic/resources/mdtest/call/union.md:
````diff
@@ -99,6 +99,47 @@ def _(flag: bool):
+def _(x: int, y: int):
+    # error: [too-many-positional-arguments]
+    Rational(x, y)
+```
````
I think this test is really just a slightly less concise version of the above test -- I don't think it adds enough value to keep it?
```suggestion
```

carljm commented on crates/ty_python_semantic/src/types/infer/builder.rs:
```diff
@@ -11753,7 +11739,10 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
+        // Validate `TypedDict` constructor calls after argument type inference.
+        // Note: We only do this for `Type::ClassLiteral` call sites today; `GenericAlias` or
+        // `type[T]` paths still use the constructor bindings fallback and do not have concrete
+        // field sets to validate against. Preserve that behavior unless we decide to tighten it.
```
Nice catch! This is just a bug, I filed https://github.com/astral-sh/ty/issues/2735

```suggestion
        // TODO this should also handle `GenericAlias` and `SubclassOf` types.
```

carljm commented on crates/ty_python_semantic/src/types/infer/builder.rs:
```diff
@@ -9813,6 +9792,9 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
             // We should see revisit this after the new constraint solver is implemented.
+            //
+            // Skip return-type assignability for constructor bindings: constructors return an
+            // instance, while narrowing here is about the callable (class) type.
```
I don't understand this comment. It seems to me that both `narrowed_ty` and the bindings return type would both be instance types here?

Removing the check in the next line doesn't fail any tests, so I'm suspicious that maybe we don't even need it. I'll check on that with a separate PR so as to check the ecosystem results separately. But I don't think we should add this comment.
```suggestion
            //
            // TODO tests pass without the `constructor_instance_type()` check on the next line; remove it?
```

Hugo-Polloli commented:
Hit a merge conflict with https://github.com/astral-sh/ruff/pull/23113 after it merged
Which I resolved so now this PR includes that broadened TypedDict constructor validation (ClassLiteral + GenericAlias + SubclassOf) alongside the constructor-bindings refactor

carljm commented:
FYI I'm making some updates to this in preparation to merge it.
